### ISSUE TITLE
Can't get MPD to display at all on my polybar

### ISSUE THREAD

**Issue by chrisco23** [polybar/polybar#2403]
I've been running polybar for a few years but I've been using Clementine with a script/module specifically for Clementine.  And using playerctl in the bar.

Now I'm running mpd w/cantata, all working fine.

But I can't get the mpd module to display anything.

**Was it working before?**
no
* Did you also experience this bug in an earlier version of polybar (yes/no/don't know)?
no
* If no, what was the last version where this worked correctly?

Here's the relevant part of my mpd.conf:
```
audio_output {
  type    "httpd"
  name    "MUB Stream"
  encoder   "vorbis"    # optional
  port    "8000"
# quality   "5.0"     # do not define if bitrate is defined
# bitrate   "128"     # do not define if quality is defined
  format    "44100:16:1"
  always_on       "yes"     # prevent MPD from disconnecting all listeners when playback is stopped.
  tags            "yes"     # httpd supports sending tags to listening streams.
}
```
And the relevant part of my polybar config:
```
[module/mpdnew]
type = internal/mpd
host = 127.0.0.1
port = 6600
interval = 1
; Available tags:
;   <label-song> (default)
;   <label-time>
;   <bar-progress>
;   <toggle> - gets replaced with <icon-(pause|play)>
;   <toggle-stop> - gets replaced with <icon-(stop|play)>
;   <icon-random>
;   <icon-repeat>
;   <icon-repeatone> (deprecated)
;   <icon-single> - Toggle playing only a single song. Replaces <icon-repeatone>
;   <icon-consume>
;   <icon-prev>
;   <icon-stop>
;   <icon-pause>
;   <icon-next>
;   <icon-seekb>
;   <icon-seekf>
;format-online = <icon-prev> <icon-seekb> <icon-stop> <toggle> <icon-seekf> <icon-next>  <icon-repeat> <icon-random>  <bar-progress> <label-time>  <label-song>
format-playing = ${self.format-online}
format-paused = ${self.format-online}
format-stopped = ${self.format-online}

; Available tags:
;   <label-offline>
; Default: ""
format-offline = <label-offline>

; Available tokens:
;   %artist%
;   %album-artist%
;   %album%
;   %date%
;   %title%
; Default: %artist% - %title%
label-song = 𝄞 %artist% - %title%

; Available tokens:
;   %elapsed%
;   %total%
; Default: %elapsed% / %total%
;label-time = %elapsed% / %total%

; Available tokens:
;   None
label-offline = 🎜 mpd is offline

; Only applies if <icon-X> is used
icon-play = ⏵
icon-pause = ⏸
icon-stop = ⏹
icon-prev = ⏮
icon-next = ⏭
icon-seekb = ⏪
icon-seekf = ⏩
icon-random = 🔀
icon-repeat = 🔁
icon-repeatone = 🔂
icon-single = 🔂
icon-consume = ✀

; Used to display the state of random/repeat/repeatone/single
; Only applies if <icon-[random|repeat|repeatone|single]> is used
toggle-on-foreground = #ff
toggle-off-foreground = #55

; Only applies if <bar-progress> is used
bar-progress-width = 45
bar-progress-indicator = |
bar-progress-fill = ─
bar-progress-empty = ─

format-online = <icon-prev> <icon-seekb> <icon-stop> <toggle> <icon-seekf> <icon-next>  <icon-repeat> <icon-random>  <bar-progress> <label-time>  <label-song>
                  
```
## Polybar Log
```
$ polybar -l info top
notice: Parsing config file: /home/mylinuxusername/.config/polybar/config
* Created ipc channel at: /tmp/polybar_mqueue.121125
* Environment var ${MONITOR} is undefined, using defined fallback value "HDMI-1"

* Loaded monitor HDMI-1 (1920x1080+1920+0)
* Bar geometry: 1920x35+1920+0; Borders: 4,4,4,4
notice: pulseaudio: using default sink alsa_output.usb-Alesis_iO4-00.analog-surround-40
* Starting application
* Starting module/i3
* Starting module/filesystem
* Starting module/openweathermap-simple
* Starting module/backlight-acpi
* Starting module/pulseaudio
* module/openweathermap-simple: Invoking shell command: "~/.config/polybar/myScripts/openweathermap-simple.sh"
* Starting module/mpdnew
* Starting module/memory
* Starting module/cpu
* Starting module/updates
* Starting module/wlan
* Starting module/bluetooth
* Starting module/battery
* Starting module/temperature
* Starting module/date
* Starting module/powermenu
* Entering event loop (thread-id=1)
* Eventqueue worker (thread-id=11)
* Configured DPI = 96x96
notice: Loaded font "RobotoMono Nerd Font:pixelsize=10" (name=RobotoMono Nerd Font, offset=0, file=/usr/share/fonts/TTF/Roboto Mono Nerd Font Complete.ttf)
notice: Loaded font "SauceCodePro Nerd Font:pixelsize=10" (name=SauceCodePro Nerd Font, offset=0, file=/usr/share/fonts/TTF/Sauce Code Pro Nerd Font Complete.ttf)
* Bar window: 0x0a00002
* Environment var ${TRAY} is undefined, using defined fallback value ""
error: Disabling tray manager (reason: Invalid position "")
* module/i3: Rebuilding cache
* module/filesystem: Rebuilding cache
* module/openweathermap-simple: Rebuilding cache
* module/backlight-acpi: Rebuilding cache
* module/pulseaudio: Rebuilding cache
* module/mpdnew: Rebuilding cache
* module/mpdnew: Hiding module since queue is empty

```

## Environment:
* WM: i3
* Distro: arch 
* Output of `polybar -vvv`:

```
$ polybar -vvv
polybar 3.5.5

Features: +alsa +curl +i3 +mpd +network(libnl) +pulseaudio +xkeyboard

X extensions: +randr (+monitors) +composite +xkb +xrm +xcursor

Build type: Release
Compiler: /usr/bin/c++
Compiler flags: -D_FORTIFY_SOURCE=2 -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -Wall -Wextra -Wpedantic -O3 -DNDEBUG
Linker flags: -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now
```

**Comment by patrick96** [comment]
> * module/mpdnew: Hiding module since queue is empty

The module is reporting that the play queue is empty. Is there anything in the mpd queue?

If so, please post your entire mpd config, the excerpt you posted only shows an audio output not where mpd is running.

**Comment by chrisco23** [comment]
Thanks for your reply.

I was listening to a playlist I created on Cantata at the time I ran those commands.

So I thought that must mean there's something in the mpd queue but maybe I am misunderstanding the mpd queue concept yet?

Here's my full mpd config:
```$ cat mpd.conf
# An example configuration file for MPD.
# Read the user manual for documentation: http://www.musicpd.org/doc/user/


# Files and directories #######################################################
#
# This setting controls the top directory which MPD will search to discover the
# available audio files and add them to the daemon's online database. This
# setting defaults to the XDG directory, otherwise the music directory will be
# be disabled and audio files will only be accepted over ipc socket (using
# file:// protocol) or streaming files over an accepted protocol.
#
#music_directory                "~/music"
music_directory                 "/mnt/mybook/Archive_2016_08/Music"
#
# This setting sets the MPD internal playlist directory. The purpose of this
# directory is storage for playlists created by MPD. The server will use
# playlist files not created by the server but only if they are in the MPD
# format. This setting defaults to playlist saving being disabled.
#
#playlist_directory             "~/.mpd/playlists"
#
# This setting sets the location of the MPD database. This file is used to
# load the database at server start up and store the database while the
# server is not up. This setting defaults to disabled which will allow
# MPD to accept files over ipc socket (using file:// protocol) or streaming
# files over an accepted protocol.
#
#db_file                        "~/.mpd/database"
#
# These settings are the locations for the daemon log files for the daemon.
# These logs are great for troubleshooting, depending on your log_level
# settings.
#
# The special value "syslog" makes MPD use the local syslog daemon. This
# setting defaults to logging to syslog.
#
#log_file                       "~/.mpd/log"
#
# This setting sets the location of the file which stores the process ID
# for use of mpd --kill and some init scripts. This setting is disabled by
# default and the pid file will not be stored.
#
#pid_file                       "~/.mpd/pid"
#
# This setting sets the location of the file which contains information about
# most variables to get MPD back into the same general shape it was in before
# it was brought down. This setting is disabled by default and the server
# state will be reset on server start up.
#
#state_file                     "~/.mpd/state"
#
# The location of the sticker database.  This is a database which
# manages dynamic information attached to songs.
#
#sticker_file                   "~/.mpd/sticker.sql"
#
###############################################################################


# General music daemon options ################################################
#
# This setting specifies the user that MPD will run as. MPD should never run as
# root and you may use this setting to make MPD change its user ID after
# initialization. This setting is disabled by default and MPD is run as the
# current user.
#
#user                           "nobody"
#
# This setting specifies the group that MPD will run as. If not specified
# primary group of user specified with "user" setting will be used (if set).
# This is useful if MPD needs to be a member of group such as "audio" to
# have permission to use sound card.
#
#group                          "nogroup"
#
# This setting sets the address for the daemon to listen on. Careful attention
# should be paid if this is assigned to anything other than the default, any.
# This setting can deny access to control of the daemon. Not effective if
# systemd socket activiation is in use.
#
# For network
#bind_to_address                "any"
#
# And for Unix Socket
#bind_to_address                "~/.mpd/socket"
#
# This setting is the TCP port that is desired for the daemon to get assigned
# to.
#
#port                           "6600"
#
# Suppress all messages below the given threshold.  Use "verbose" for
# troubleshooting. Available setting arguments are "notice", "info", "verbose",
# "warning" and "error".
#
#log_level                      "notice"
#
# Setting "restore_paused" to "yes" puts MPD into pause mode instead
# of starting playback after startup.
#
#restore_paused "no"
#
# This setting enables MPD to create playlists in a format usable by other
# music players.
#
#save_absolute_paths_in_playlists       "no"
#
# This setting defines a list of tag types that will be extracted during the
# audio file discovery process. The complete list of possible values can be
# found in the user manual.
#metadata_to_use        "artist,album,title,track,name,genre,date,composer,performer,disc"
#
# This example just enables the "comment" tag without disabling all
# the other supported tags:
#metadata_to_use "+comment"
#
# This setting enables automatic update of MPD's database when files in
# music_directory are changed.
#
#auto_update    "yes"
#
# Limit the depth of the directories being watched, 0 means only watch
# the music directory itself.  There is no limit by default.
#
#auto_update_depth "3"
#
###############################################################################


# Symbolic link behavior ######################################################
#
# If this setting is set to "yes", MPD will discover audio files by following
# symbolic links outside of the configured music_directory.
#
#follow_outside_symlinks        "yes"
#
# If this setting is set to "yes", MPD will discover audio files by following
# symbolic links inside of the configured music_directory.
#
#follow_inside_symlinks         "yes"
#
###############################################################################


# Zeroconf / Avahi Service Discovery ##########################################
#
# If this setting is set to "yes", service information will be published with
# Zeroconf / Avahi.
#
#zeroconf_enabled               "yes"
#
# The argument to this setting will be the Zeroconf / Avahi unique name for
# this MPD server on the network. %h will be replaced with the hostname.
#
#zeroconf_name                  "Music Player @ %h"
#
###############################################################################


# Permissions #################################################################
#
# If this setting is set, MPD will require password authorization. The password
# setting can be specified multiple times for different password profiles.
#
#password                        "password@read,add,control,admin"
#
# This setting specifies the permissions a user has who has not yet logged in.
#
#default_permissions             "read,add,control,admin"
#
###############################################################################


# Database #######################################################################
#
# An example of a database section instead of the old 'db_file' setting.
# It enables mounting other storages into the music directory.
#
#database {
#       plugin "simple"
#       path "~/.local/share/mpd/db
#       cache_directory "~/.local/share/mpd/cache"
#}
#
# An example of database config for a sattelite setup
#
#music_directory "nfs://fileserver.local/srv/mp3"
#database {
#       plugin "proxy"
#       host "other.mpd.host"
#       port "6600"
#}

# Input #######################################################################
#
input {
        plugin "curl"
#       proxy "proxy.isp.com:8080"
#       proxy_user "user"
#       proxy_password "password"
}

#
###############################################################################

# Audio Output ################################################################
#
# MPD supports various audio output types, as well as playing through multiple
# audio outputs at the same time, through multiple audio_output settings
# blocks. Setting this block is optional, though the server will only attempt
# autodetection for one sound card.
#
# An example of an ALSA output:
#
#audio_output {
#       type            "alsa"
#       name            "My ALSA Device"
##      device          "hw:0,0"        # optional
##      mixer_type      "hardware"      # optional
##      mixer_device    "default"       # optional
##      mixer_control   "PCM"           # optional
##      mixer_index     "0"             # optional
#}
#
# An example of an OSS output:
#
#audio_output {
#       type            "oss"
#       name            "My OSS Device"
##      device          "/dev/dsp"      # optional
##      mixer_type      "hardware"      # optional
##      mixer_device    "/dev/mixer"    # optional
##      mixer_control   "PCM"           # optional
#}
#
# An example of a shout output (for streaming to Icecast):
#
#audio_output {
#       type            "shout"
#       encoder         "vorbis"                # optional
#       name            "My Shout Stream"
#       host            "localhost"
#       port            "8000"
#       mount           "/mpd.ogg"
#       password        "hackme"
#       quality         "5.0"
#       bitrate         "128"
#       format          "44100:16:1"
##      protocol        "icecast2"              # optional
##      user            "source"                # optional
##      description     "My Stream Description" # optional
##      url             "http://example.com"    # optional
##      genre           "jazz"                  # optional
##      public          "no"                    # optional
##      timeout         "2"                     # optional
##      mixer_type      "software"              # optional
#}
#
# An example of a recorder output:
#
#audio_output {
#       type            "recorder"
#       name            "My recorder"
#       encoder         "vorbis"                # optional, vorbis or lame
#       path            "/var/lib/mpd/recorder/mpd.ogg"
##      quality         "5.0"                   # do not define if bitrate is defined
#       bitrate         "128"                   # do not define if quality is defined
#       format          "44100:16:1"
#}
#
# An example of a httpd output (built-in HTTP streaming server):
#
#audio_output {
#       type            "httpd"
#       name            "My HTTP Stream"
#       encoder         "vorbis"                # optional, vorbis or lame
#       port            "8000"
#       bind_to_address "0.0.0.0"               # optional, IPv4 or IPv6
##      quality         "5.0"                   # do not define if bitrate is defined
#       bitrate         "128"                   # do not define if quality is defined
#       format          "44100:16:1"
#       max_clients     "0"                     # optional 0=no limit
#}
#
# An example of a pulseaudio output (streaming to a remote pulseaudio server)
#
#audio_output {
#       type            "pulse"
#       name            "My Pulse Output"
##      server          "remote_server"         # optional
##      sink            "remote_server_sink"    # optional
##      media_role      "media_role"            #optional
#}
#
# An example of a winmm output (Windows multimedia API).
#
#audio_output {
#       type            "winmm"
#       name            "My WinMM output"
##      device          "Digital Audio (S/PDIF) (High Definition Audio Device)" # optional
#               or
##      device          "0"             # optional
##      mixer_type      "hardware"      # optional
#}
#
# An example of a wasapi output (Windows multimedia API).
#
#audio_output {
#       type            "wasapi"
#       name            "My WASAPI output"
##      device          "Digital Audio (S/PDIF) (High Definition Audio Device)" # optional
#               or
##      device          "0"             # optional
## Exclusive mode blocks all other audio source, and get best audio quality without resampling.
##      exclusive       "no"            # optional
## Enumerate all devices in log.
##      enumerate       "no"            # optional
#}
#
# An example of an openal output.
#
#audio_output {
#       type            "openal"
#       name            "My OpenAL output"
##      device          "Digital Audio (S/PDIF) (High Definition Audio Device)" # optional
#}
#
# An example of an sndio output.
#
#audio_output {
#       type            "sndio"
#       name            "sndio output"
#       mixer_type      "hardware"
#}
#
# An example of an OS X output:
#
#audio_output {
#       type            "osx"
#       name            "My OS X Device"
##      device          "Built-in Output"       # optional
##      channel_map      "-1,-1,0,1"    # optional
#}
#
## Example "pipe" output:
#
#audio_output {
#       type            "pipe"
#       name            "my pipe"
#       command         "aplay -f cd 2>/dev/null"
## Or if you're want to use AudioCompress
#       command         "AudioCompress -m | aplay -f cd 2>/dev/null"
## Or to send raw PCM stream through PCM:
#       command         "nc example.org 8765"
#       format          "44100:16:2"
#}
#
## An example of a null output (for no audio output):
#
#audio_output {
#       type            "null"
#       name            "My Null Output"
#       mixer_type      "none"                  # optional
#}
#
###############################################################################


# Normalization automatic volume adjustments ##################################
#
# This setting specifies the type of ReplayGain to use. This setting can have
# the argument "off", "album", "track" or "auto". "auto" is a special mode that
# chooses between "track" and "album" depending on the current state of
# random playback. If random playback is enabled then "track" mode is used.
# See <https://wiki.hydrogenaud.io/index.php?title=Replaygain> for
# more details about ReplayGain.
# This setting is off by default.
#
#replaygain                     "album"
#
# This setting sets the pre-amp used for files that have ReplayGain tags. By
# default this setting is disabled.
#
#replaygain_preamp              "0"
#
# This setting sets the pre-amp used for files that do NOT have ReplayGain tags.
# By default this setting is disabled.
#
#replaygain_missing_preamp      "0"
#
# This setting enables or disables ReplayGain limiting.
# MPD calculates actual amplification based on the ReplayGain tags
# and replaygain_preamp / replaygain_missing_preamp setting.
# If replaygain_limit is enabled MPD will never amplify audio signal
# above its original level. If replaygain_limit is disabled such amplification
# might occur. By default this setting is enabled.
#
#replaygain_limit               "yes"
#
# This setting enables on-the-fly normalization volume adjustment. This will
# result in the volume of all playing audio to be adjusted so the output has
# equal "loudness". This setting is disabled by default.
#
#volume_normalization           "no"
#
###############################################################################

# Character Encoding ##########################################################
#
# If file or directory names do not display correctly for your locale then you
# may need to modify this setting.
#
#filesystem_charset             "UTF-8"
#
###############################################################################

audio_output {
        type            "httpd"
        name            "MUB Stream"
        encoder         "vorbis"                # optional
        port            "8000"
#       quality         "5.0"                   # do not define if bitrate is defined
#       bitrate         "128"                   # do not define if quality is defined
        format          "44100:16:1"
        always_on       "yes"                   # prevent MPD from disconnecting all listeners when playback is stopped.
        tags            "yes"                   # httpd supports sending tags to listening streams.
}
```

**Comment by patrick96** [comment]
I think there is one of two things happening here:

Either, there are two mpd instances running and polybar connects to a different one than cantata.

Or cantata maintains its own queue and leaves the mpd queue empty. This would be kind of weird because I don't think mpd can even play anything if the queue is empty, at least the current song must be in the queue.

To check whether the second one is the case, can you run the commands `mpc` and `mpc playlist`, the first command should show you the currently playing song and the second should list all songs in the queue.

**Comment by chrisco23** [comment]
Thanks.  I did some fiddling but now I have another strange situation, although your idea of two streams might be a clue here.

At some point I lost the whole mpd database, then I had cantata just showing a single album (and I have tons), no idea the significance of the one album.

But somehow the MPD controls are showing partially.  Now I'm back to being able to play my music at least in cantata, but somehow the mpd module is still showing a song from that one aforementioned album.  There is a progress bar (at 0:00 which makes sense since I'm not actually playing that).  I'm seeing 30 chars or so of format %artist% - %song% next to it.

`$ mpc
45 Grave - Insurance From God
[paused]  #1/14   0:00/5:06 (0%)
volume: n/a   repeat: off   random: off   single: off   consume: off
ERROR: Failed to enable output "MUB Stream" (httpd); Failed to bind to '[::]:6600'; Failed to bind socket: Address alre
ady in use
08:50 AM:fugbert@xpschris:~:
$ mpc playlist
45 Grave - Insurance From God
45 Grave - Evil
45 Grave - Partytime
45 Grave - Violent World
45 Grave - Slice O' Life
45 Grave - Phantoms
45 Grave - Dream Hits II
45 Grave - Bad Love
45 Grave - Surf Bat
45 Grave - Riboflavin
45 Grave - Procession
45 Grave - 45 Grave
45 Grave - School's Out
45 Grave - Partytime (Single Version)`

I had never even tried to play this other than that for a moment last night, it was the only album showing up in cantata, so I had attempted to play the first track.

Sorry no linefeeds there but hopefully the mpc output is good.  The mess is because I run i3 but with a few KDE apps, and I'm cleaning those out (konsole and dolphin mainly) and trying to deal with a new terminal emulator (trying alacritty for the moment).

**Comment by patrick96** [comment]
I assume polybar is consistent with the output of the `mpc` command while cantata shows completely different songs?

I guess, it is most likely that there are two mpd instances. Maybe one is running as root and one as a normal user?

I'm closing this, since it isn't related to polybar.

**Comment by chrisco23** [comment]
I just returned to this and figured it out.  Sorry for the trouble.

Since I didn't see much displayed at all I assumed it was the MPD module, but I found that there was a gap in my understanding of polybar fonts.  Specifically I had to use gucharmap, test-fonts.pl, and install 3 fonts (I chose ttf-symbola, noto-fonts-emoji, and siji-git) to get everything displaying, after which everything seems to be working.
